tmt/utils.py

```diff
@@ -289,6 +289,53 @@ def get_output(self) -> str:
 
 CommonDerivedType = TypeVar('CommonDerivedType', bound='Common')
 
+# TODO: once mypy learns to verify LiteralString, we should really, really enforce
+# the type of each and every command element. It would also make the analysis stronger
+# as it would be able to detect places where text blobs containing shell scripts are
+# injected into commands that would not be handled executed under shell.
+#
+# So, something like this would be one of the options.
+#
+# QuotedString = NewType('QuotedString', str)
+# ShellScript = NewType("ShellScript", str)
+#
+# Commands - executed without `shell=True`/`bash -c ...` - consist of either literal strings
+# or strings:
+#
+#   CommandElement = Union[LiteralString, str]
+#   Command = List[CommandElement]
+#
+# Commands to be execuetd with `shell=True`/`bash -c` can consist of literal strings or
+# correctly escaped strings:
+#
+#   ShellCommandElement = Union[LiteralString, QuotedString]
+#   ShellCommand = List[ShellCommandElement]
+#
+# And we'd have nice helpers to handle conversions (mypy would not let us combine literal strings
+# and values given to tmt via `--script` options & like):
+#
+# def quote_string(s: str) -> QuotedString:
+#     return QuotedString(shlex.quote(s))
+#
+# Bad:  run(['cat'] + [             path  for path in self.opt('paths-to-list')], shell=True)
+# Good: run(['cat'] + [quote_string(path) for path in self.opt('paths-to-list')], shell=True)
+#
+# Until that, at least make a distinction between a command-like and shell-like strings.
+ShellScript = NewType('ShellScript', str)
+_CommandElement = str
+Command = List[_CommandElement]
+
+
+def command_to_command_element(command: Command) -> _CommandElement:
```

Comment:
So this takes a `Command` which is a list of `_CommandElement`s and converts it into a single `_CommandElement`? Sounds a be weird to me. Shouldn't the return value be a `str`? Have I completely missed the purpose of the `_CommandElement`?

Reply:
This exists to "squeeze" command into a single shell command token, e.g. `['ls', '-al', '/']` => `['bash', '-c', command_to_command_element(['ls', '-al', '/'])]`.

The name sucks... But the purpose should be really this, to take a command that's supposed to be part of a command line, like SSH options for Ansible: `'--ssh-common-args', command_to_command_element(self._ssh_command())`.

The return value should really be `_CommandElement` (even if renamed...), i.e. a single token of a command line. Yes, `_CommandElement` is a string, the name gives mypy and us a chance to follow "this is part of a command!" strings through the code, and when mypy gets smarter, prevent conflicts with "ordinary" strings. Therefore these functions stick to `_CommandElement` even if it's a `str` alias today.

Reply:
Clarified on the call. Thanks!